### PR TITLE
Normalize hr, cadence and power averages

### PR DESCRIPTION
**Pull request by rgmf**
**Describe the pull request**
It computes time weighted heart rate, cadence and power averages. So the more time it takes the more important is.
Also, it computes heart rate and cadence maximum.
Finally, it ignores pause time for whatever reason (SEGMENT_START_MANUAL, SEGMENT_START_AUTOMATIC).

For a track with about 7000 trackpints the query takes about 140ms.

**Link to the the issue**
#601 

**License agreement**
By opening this pull request, you are providing your contribution under the _Apache License 2.0_ (see [LICENSE.md](LICENSE.md)).

### CODE REVIEW

**Comment by rgmf** [comment]
@dennisguse if you think it's okay then I'll write tests to ensure all this is working well (we cannot trust on manual tests for this).

File: src/main/java/de/dennisguse/opentracks/content/data/TrackPointsColumns.java [hunk 1]
```diff
@@ -31,6 +31,7 @@
     String TABLE_NAME = "trackpoints";
     Uri CONTENT_URI_BY_ID = Uri.parse(ContentProviderUtils.CONTENT_BASE_URI + "/" + TABLE_NAME);
     Uri CONTENT_URI_BY_TRACKID = Uri.parse(ContentProviderUtils.CONTENT_BASE_URI + "/" + TABLE_NAME + "/trackid");
+    Uri CONTENT_URI_SENSOR_STATS = Uri.parse(ContentProviderUtils.CONTENT_BASE_URI + "/" + TABLE_NAME + "/sensorstats");
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/content/data/TrackPointsColumns.java:34]: I would consider this computation rather part of the track than of the trackpoints.
Technically, you are completely right as the data is derived from the trackpoints.
However, in the track-table, we have all the other statistics - only difference is that those are precomputed.

File: src/main/java/de/dennisguse/opentracks/content/provider/CustomContentProvider.java [hunk 2]
```diff
@@ -356,11 +362,46 @@ private Uri insertMarker(Uri url, ContentValues contentValues) {
         throw new SQLException("Failed to insert a marker " + url);
     }
 
+    /**
+     * Return the string representing the query that compute sensor stats from trackpoints table.
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/content/provider/CustomContentProvider.java:366]: * Return the string representing the query that compute sensor stats from trackpoints table.
* It computes the average for heart rate, cadence, and power (duration-based average) and the maximum for heart rate and cadence.

File: src/main/java/de/dennisguse/opentracks/content/provider/CustomContentProvider.java [hunk 3]
```diff
@@ -356,11 +362,46 @@ private Uri insertMarker(Uri url, ContentValues contentValues) {
...
+     * It computes time weighted heart rate, cadence and power averages. So the more time it takes the more important is.
+     * Also, it computes heart rate and cadence maximum.
+     * Finally, it ignores pause time for whatever reason (SEGMENT_START_MANUAL, SEGMENT_START_AUTOMATIC).
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/content/provider/CustomContentProvider.java:369]: I would argue that a START_AUTOMATIC should be considered as it only denotes a gap in the GPS track (due to fast movement, low reception etc).
Example: 
If I use a cycling sensor (distance/speed), this sensor data is stored for all trackpoints of type TRACKPOINT, START_AUTOMATIC, and (I guess even) END_MANUAL. 
If I am cycling in the mountains, GPS reception is likely to be varying a lot and this will likely result in START_AUTOMATIC (like going downhill with a racing bike at 60km/h).
And in my opinion, these should be part of the AVG/MAX computation.

In data:
t=0, START_MANUAL
t=10, TRACKPOINT, speed=5
t=20, TRACKPOINT, speed=10
t=25, START_AUTOMATIC speed=60
t=27, START_AUTOMATIC speed=60
t=30, TRACKPOINT, speed=15
t=30, END_MANUAL, speed=15

I was thinking about this while working on the "distance from sensor" task and for the overall sum to match, we need to consider all trackpoints that have distance (relative to previous).
So, basically everything except START_MANUAL, which should not have any sensor data (as no connection to sensor is yet established) and it has not duration to previous trackpoint.

**Review by dennisguse — CHANGES_REQUESTED** [review]
Code looks good; some comments.
(almost good to go!)

File: src/main/java/de/dennisguse/opentracks/content/provider/CustomContentProvider.java (hunk 3 above)
Comment (rgmf) [reply src/main/java/de/dennisguse/opentracks/content/provider/CustomContentProvider.java:369]: Umm, okay, I understand. I was thinking on those moments when you take a rest in your route and forget pause the activity. I guess we cannot do anything here.

So, yes, you are right.

File: src/main/java/de/dennisguse/opentracks/content/provider/CustomContentProvider.java [hunk 4]
```diff
@@ -356,13 +362,47 @@ private Uri insertMarker(Uri url, ContentValues contentValues) {
         throw new SQLException("Failed to insert a marker " + url);
     }
 
+    /**
+     * Return the string representing the query that compute sensor stats from trackpoints table.
+     * It computes the average for heart rate, cadence, and power (duration-based average) and the maximum for heart rate and cadence.
+     * Finally, it ignores manual pause (SEGMENT_START_MANUAL).
+     */
+    private final String getSensorStatsQuery() {
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/content/provider/CustomContentProvider.java:370]: Can be a property aka private final String SENSOR_STATS_QUERY = ""

**Review by dennisguse — APPROVED** [review]
Code looks good!

File: src/main/java/de/dennisguse/opentracks/content/provider/CustomContentProvider.java (hunk 4 above)
Comment (rgmf) [reply src/main/java/de/dennisguse/opentracks/content/provider/CustomContentProvider.java]: Oops!!! :see_no_evil: I read it quickly... Fixed!

**Comment by rgmf** [comment]
Ready to go...

I've added tests to check query is computing averages well.

**Review by dennisguse — APPROVED** [review]
Looks good!
@rgmf Merge at your discretion.
